### PR TITLE
Fix in(x, r::Range{<:Integer}) when x is not numeric

### PR DESCRIPTION
**Pull request by nalimilan**
Previously, checking that a string was in() a container would work for Array
but not for Range.

Another (possibly cleaner) approach would be to restrict the current definitions to `x::Number`, and introduce fallbacks returning `false`.

### CODE REVIEW

**Review by Sacha0 — COMMENTED** [review]
> Another (possibly cleaner) approach would be to restrict the current definitions to x::Number, and introduce fallbacks returning false.

The alternative you propose indeed sounds cleaner and clearer :). (Knowing nothing about `applicable`, I wonder whether the added `applicable` calls and associated branches impact performance or are optimized away?) Best!

**Comment by nalimilan** [comment]
> The alternative you propose indeed sounds cleaner and clearer :).

Yeah, maybe. I was hesitant because of the possibility of non-`Number` types supporting ranges, like `Date`. Not sure.

> (Knowing nothing about applicable, I wonder whether the added applicable calls and associated branches impact performance or are optimized away?) Best!

It should be resolved at compile time like `method_exists`, so there should be no runtime cost.

**Review by TotalVerb — COMMENTED** [review]
I also like the alternative better than this. This seems brittle for whatever reason.

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: Tangentially, does this branch potentially introduce type instability?

**Comment by Sacha0** [comment]
> I was hesitant because of the possibility of non-Number types supporting ranges, like Date.

Agreed, tricky that. I suppose this method is an instance where the ability to dispatch on provision of the necessary arithmetic operations would be useful.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: Why? It doesn't define any variable, and makes the function essentially equivalent to `return false` when the condition is false.

Comment (Sacha0) [inline base/range.jl]: The (preexisting) branch in the `n = ...` line, rather than the new `applicable(...` line :). To expand, down one path `n = 1`, whereas down the other `n = round(Integer, ...`. If the `round` yields some non-`Int` `Integer`, `n`'s type would depend on the branch? Best!

**Comment by nalimilan** [comment]
I think this illustrates the fact that the interfaces are not defined very clearly here. Ranges can be defined on any type which supports addition and subtraction. The ideal solution for this would be a trait.

I would be happy to use the cleaner `Number` approach, but that will force anybody willing to support ranges with a non-`Number` custom type to copy the very same definition of `in`, but with a different signature. Is that really what we want?

**Comment by Sacha0** [comment]
> I would be happy to use the cleaner Number approach, but that will force anybody willing to support ranges with a non-Number custom type to copy the very same definition of in, but with a different signature. Is that really what we want?

Perhaps for now keep the `applicable(-, ...` in the first method definition (where traits would be ideal), and in the second method definition (presently with `applicable(isinteger, ...`) use dispatch? (In the second method definition I see no disadvantage to using dispatch, given that checking `applicable(isinteger, ...` should be equivalent to dispatching on `x::Integer`?)

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: Mmm. You're right that when the condition is true, the rest of the body isn't optimized out as I would have expected. So `n`  still exists and is inferred as `Any`; though the return value is correctly inferred as `Bool`.

More fundamentally, `applicable`/`method_exists` does not appear to be resolved at compile time. That's because https://github.com/JuliaLang/julia/pull/16422 hasn't been merged (I had forgotten about this). So I have no solution for now...

**Comment by nalimilan** [comment]
Yes, for the second definition dispatch would work equally well. The first one is more tricky (especially if we can't get `applicable` to be resolved at compile time).

**Comment by Sacha0** [comment]
> Yes, for the second definition dispatch would work equally well. The first one is more tricky (especially if we can't get applicable to be resolved at compile time).

In that case, perhaps for now nix the change to the first method definition, use dispatch in the second, merge the result, and revisit the first method definition when relevant capabilities advance? :)

**Comment by JeffBezanson** [comment]
How did this come up? We do just give errors in some cases like this, for example `searchsortedfirst([1,2,3], "")` is similar.

> applicable(isinteger, ... should be equivalent to dispatching on x::Integer

`isinteger(2.0)` is also true, so should probably dispatch on `::Number`.

**Comment by nalimilan** [comment]
> How did this come up? We do just give errors in some cases like this, for example searchsortedfirst([1,2,3], "") is similar.

I discovered this when making a mistake (as often with corner cases like this). It's not a major issue, but I find it hard to justify the inconsistency with `Array`. A relatively credible use case would be this:
```julia
julia> Nullable() in [1, 2]
false

julia> Nullable() in 1:2
ERROR: MethodError: no method matching isinteger(::Nullable{Union{}})

```
This would become even more credible if we replace `Nullable` with `Union{T, Null}`, with `==` defined so that `null in [1, 2, null]` would be `true` (currently that's a `NullException`).

**Comment by nalimilan** [comment]
I've dropped the problematic change for now. Though I still would be interested in suggestions about how to fix this for the non-integer case.

**Comment by JeffBezanson** [comment]
Makes sense. Yes, it's a good question. Maybe we could dispatch as:

```
in(x::Union{Number,T}, r::Range{T}) where T
```

**Comment by nalimilan** [comment]
But that wouldn't change anything for `"a" in 1:3`, right?

**Comment by JeffBezanson** [comment]
If that definition doesn't match, we'd fall back to iterating over the range and comparing every element, or we could add a fallback that returns `false`.

**Comment by nalimilan** [comment]
Clever. Indeed with https://github.com/JuliaLang/julia/issues/21256 the fallback based on `any` should be optimized at compile time if `==` isn't defined for a pair of immutable types (and falls back to `===`).

Looking at this in more detail, it turns out we need to restrict the signature to `Real`, not `Number`, since `isinteger(Complex(1, 0)) == true` and yet `<` isn't supported. So I've added a specific method for `Complex` too. Also, the `in(x::Union{Number,T}, r::Range{T}) where T` trick doesn't work since  `-(::T, Number)` is not always implemented, only `-(::T, ::T)` may be supported (e.g. for `Date`).

At least now the abstraction is correctly respected: the fallback is used for types for which we're not sure an optimized computation is possible given the interface.

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: Could this signature simplify to `in(x::Real, r::Range{<:Integer})`?

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: Perhaps create a helper function to avoid replication of the body of this / the below method?

File: base/complex.jl
Comment (Sacha0) [inline base/complex.jl]: Could this signature simplify to `in(x::Complex, r::Range{<:Integer})`?

File: base/complex.jl
Comment (Sacha0) [inline base/complex.jl]: The equivalent method below for `x::Real` converts `x` to type `T` prior to the `mod`. Should that be done here as well?

**Comment by Sacha0** [comment]
Could the separate methods for `z::Complex` and `x::Real` not be fused into a single method for `x::Number` that includes the `real` calls from the `z::Complex` method? For `x::Real`, those calls should be noops? Best!

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: Yes, I kept it because `T` is used.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: Not sure, are these two lines worth the additional complexity?

File: base/complex.jl
Comment (nalimilan) [inline base/complex.jl]: Good question. I'll add it back since that must have been added for a reason (maybe to prevent overflow), but I don't know.

**Comment by nalimilan** [comment]
> Could the separate methods for z::Complex and x::Real not be fused into a single method for x::Number that includes the real calls from the z::Complex method? For x::Real, those calls should be noops? Best!

Unfortunately, `Complex` isn't defined at that point in the bootstrap.

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: A single implementation in a helper function strikes me as less complex :). Moreover, guards against changes being made in one place but not the other.

Comment (nalimilan) [inline base/range.jl]: Given that the definitions are so close, I doubt it. Also it turns out it's harder than it seems, since their signatures are different: `T` doesn't need to be the same for both arguments with the `Real` method.

**Comment by Sacha0** [comment]
> Unfortunately, Complex isn't defined at that point in the bootstrap.

Hm, pity that. DRYing those methods would be nice, as evidenced by the inadvertent conversion discrepancy creeping in above.

File: base/complex.jl
Comment (Sacha0) [inline base/complex.jl]: Perhaps in case only `mod(::T, ::T)` (rather than also `mod(::typeof(x), ::T)`) exists?

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: It turns out that method generates ambiguities with other methods defined in this file if I change `<:T` to `T` (as it should be). So we need to decide what choice is the most painful: requiring people to define this method themselves when they implement a custom non-`Real` type supporting ranges; or define this generic fallback here (current state of the PR), but require people to fix ambiguities when they need to define a custom method like this one.

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: > Given that the definitions are so close, I doubt it.

Even in the relatively short time I've been reviewing actively, IIRC I've seen such happen at least once :).

> Also it turns out it's harder than it seems, since their signatures are different: T doesn't need to be the same for both arguments with the Real method.

Not certain I follow? Would e.g. the following not work?
```julia
in(x::Real, r::Range{<:Real}) = _therecanbeonlyone(x, r)
in(x::T, r::Range{<:T}) where {T} = _therecanbeonlyone(x, r)
function _therecanbeonlyone(x, r)
    n = step(r) == 0 ? 1 : round(Integer, (x - first(r))/step(r)) + 1
    n >= 1 && n <= length(r) && r[n] == x
end
```
Best! :)

**Comment by nalimilan** [comment]
Actually I found a very simple way to DRY the `Complex` definition: define it as `
in(x::Complex, r::Range{<:Real}) = isreal(x) && real(x) in r`. This is even better than the previous one, which was limited to integers.

So the only remaining issue is the ambiguities `in(x::T, r::Range{T})` introduces.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: Of course! I don't know why I wanted to use a loop with `@eval`...

File: base/complex.jl
Comment (Sacha0) [inline base/complex.jl]: Beautiful! :)

File: base/range.jl
Comment (Sacha0) [inline base/range.jl]: The tangential discussion re. the potential type instability in this line got buried above. Any thoughts re. addressing this type instability while you're performing surgery? :)

Comment (nalimilan) [inline base/range.jl]: Ah, right. I'll fix that once we have sorted out the other issues since the discussion is already long. I wonder what's the best way of fixing this. Maybe compute the second operand unconditionnaly and call `one` on that? That would also avoid a branch if we use `ifelse`.

Comment (Sacha0) [inline base/range.jl]: > Maybe compute the second operand unconditionally and call one on that?

I imagine the branch exists to avoid the potential division by zero (`step(r)`) in computing the second operand (and potential downstream issues)?

Perhaps continuing to branch on `step(r) == 0`, but eliminating `n` in the `step(r) == 0` case would work well? For example, perhaps something along the lines of
```julia
function _in_range(x, r::Range)
    if step(r) == 0
        isempty(r) ? false : first(r) == x
    else
        n = round(Integer, (x - first(r)) / step(r)) + 1
        n >= 1 && n <= length(r) && r[n] == x
    end
end
```
Thoughts?

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: @JeffBezanson What do you think about using `Range{<:T}` rather than `Range{T}` just to prevent ambiguities here (see [my previous comment](https://github.com/JuliaLang/julia/pull/21728#discussion_r115790401))? This feels like a hack to me and I'd be inclined to remove that method even if that forces people implementing non-`Real` types which support ranges to define it manually.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: I've added a commit doing this.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: I've changed `<:T` to `T`  and it doesn't seem to generate ambiguities now, so let's go with that.

**Comment by nalimilan** [comment]
@Sacha0 Good to go now?

**Comment by Sacha0** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Review by Sacha0 — APPROVED** [review]
lgtm modulo nanosoldier's signoff! :)

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21728#issuecomment-308804455) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/df289e96c9f9bf7e47962c11342ad490b6f39b0c/cf5ca9a_vs_52069d7/report.md). cc @jrevels

**Comment by nalimilan** [comment]
The only regression is in `broadcast`, and it doesn't seem related (though it's hard to tell).

**Comment by Sacha0** [comment]
Imagine so, and @nanosoldier `runbenchmarks("broadcast" && "dotop" , vs = ":master")` can verify :).

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21728#issuecomment-308857958) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/22cbb9fc660f50fc21337b7ba623d109916b11e9/7e190bf_vs_1a3391f/report.md). cc @jrevels

File: base/range.jl
Comment (tkelman) [inline base/range.jl]: where is the fallback for `in(x::Any, r::Range{<:Integer})` ?

Comment (nalimilan) [inline base/range.jl]: That's the generic iterator fallback: `in(x, itr) = any(y -> y == x, itr)`. With https://github.com/JuliaLang/julia/issues/21256, it will be optimized out at compile time when types cannot possibly be equal.

Comment (tkelman) [inline base/range.jl]: that's an open issue not a closed PR, so we don't have that yet? are any cases that would dispatch to this signature tracked in the benchmarks?

Comment (nalimilan) [inline base/range.jl]: Indeed it's not merged yet, but that code path failed previously (which is the point of this PR), so I don't think performance matters.

Comment (Sacha0) [inline base/range.jl]: @tkelman, does @nalimilan's response address your concerns? If so this PR seems in good shape? :)

Comment (tkelman) [inline base/range.jl]: What failed about that code path? It seems to me that the new code is more restrictive than the previous version

Comment (nalimilan) [inline base/range.jl]: It's more restrictive in the sense that the fast path is only used for types for which we're it will work (i.e. `Real`, for which `isinteger` and arithmetic operations must be defined). But it's not more restrictive in general thanks to the `in(x, itr)` fallback. As noted in the description, previously checking that a non-numeric type was `in` an integer range would fail, which is quite "restrictive" (e.g. `"a" in 1:10` wouldn't work).

Comment (Sacha0) [inline base/range.jl]: Seems in shape then @tkelman? :)

**Comment by nalimilan** [comment]
@tkelman Good to go then?

File: base/range.jl
Comment (tkelman) [inline base/range.jl]: That seems considerably more expensive for inputs that are not `Real` or `Complex`.

Comment (nalimilan) [inline base/range.jl]: Yes, these types will have to provide their optimized method. I couldn't find any other approach which would always be correct. Do you have suggestions?

Comment (tkelman) [inline base/range.jl]: do we have any wider unions / abstract types that would serve as a standin for `isinteger` and `<=` existing?

Comment (nalimilan) [inline base/range.jl]: Not that I know of. FWIW, `isinteger` is only defined for `Real`, `Complex` and `Irrational` in Base. `Number` is too broad for `<=`. This sounds like a typical case where a trait would be useful (once we have traits).

BTW, when you say:
> That seems considerably more expensive for inputs that are not Real or Complex.

To be precise, `in` will still be as fast as before for any arbitrary type. What will be slow is calling `in` with an element of a custom non-`Real` type different from the range's eltype. This is probably not very common.

Comment (tkelman) [inline base/range.jl]: integer ranges are pretty common. that is exactly the case I was pointing out. would any of the traits added for recent range work be appropriate to leverage here?

Comment (nalimilan) [inline base/range.jl]: Integer ranges are still covered by the fast path with this PR since they are `Real` (see the full diff). I don't think the new range traits would make any difference: we're dealing with a much simpler question here, "does `T` support arithmetic?".

Comment (tkelman) [inline base/range.jl]: it's testing if a custom type is in an integer range that now uses the slow path

Comment (nalimilan) [inline base/range.jl]: Yes, but as I said before that it wouldn't throw an error, so that's rather an improvement than a regression.

Comment (tkelman) [inline base/range.jl]: it would throw an error unless it implements comparison and isinteger. the slowdown wouldn't be all that visible in that case

Comment (nalimilan) [inline base/range.jl]: So you're thinking about a type which would implement `isinteger` and `<`, would be commonly checked for `in` an integer range, but wouldn't be a `Real`? Do you have examples in mind?

I'm fine with raising a deprecation warning if we want to protect people from silent slowdowns.

Comment (tkelman) [inline base/range.jl]: The most likely example I can think of would be someone forgetting to subtype Real.

Comment (nalimilan) [inline base/range.jl]: So would you prefer printing a deprecation warning (and later a `MethodError`)?

Comment (nalimilan) [inline base/range.jl]: @tkelman Bump.

Comment (tkelman) [inline base/range.jl]: a deprecation would be a bit strange but would at least make the change visible. let me take one more look at this.

File: base/irrationals.jl
Comment (rfourquet) [inline base/irrationals.jl]: Already defined two lines above?

File: base/range.jl
Comment (rfourquet) [inline base/range.jl]: I understand that this makes the assumption that in this case, `-(::T, ::T)` at least returns a `Real`? as well as step `Range(::T)`?

**Review by rfourquet — APPROVED** [review]

File: base/range.jl
Comment (rfourquet) [inline base/range.jl]: If we really want to take this case into account, where `isinteger` is defined but the type isn't a subtype of `Real`, it would be possible to dispatch on the `isinteger` trait, no? But it appears overkill to me.

Comment (tkelman) [inline base/range.jl]: which trait? isinteger is value dependent

Comment (rfourquet) [inline base/range.jl]: Well, let's say a runtime trait: 
```
in(x, r::Range{T}) where {T<:Integer} = isinteger(x) ?
    _fast_real_method_below(x, r) :
    (x isa Real ? false : _slow_generic_iterator_method(x, r))
```
where `_fast_real_method_below` refers to `in(x::Real, r::Range{T}) where {T<:Integer}` defined below, where the `isinteger` test can be removed.
This would not work well for e.g. `1+0im`, which is `isinteger` but not `Real`; then in `_fast_real_method_below`, the use of `x` can be replaced by `Integer(x)`.

**Comment by StefanKarpinski** [comment]
We've got two approvals and green tests. Time to merge?

File: base/irrationals.jl
Comment (tkelman) [inline base/irrationals.jl]: definitely redundant, should be removed

Comment (nalimilan) [inline base/irrationals.jl]: Right, it's been added in another PR since I opened this one. I've just removed it.

File: base/range.jl
Comment (nalimilan) [inline base/range.jl]: The assumption is that `-(::T, ::T)` returns something which can be divided by the range's step, and the result be rounded to `Integer`. But that shouldn't be an issue since this method will only be called if a `Range{T}` has been constructed, which means these operations are supported (or should be). The problem with the current implementation is that arithmetic operations could be called on any type which isn't supposed to support them at all.

File: base/range.jl
Comment (rfourquet) [inline base/range.jl]: @tkelman are you fine with the state of this PR on this point?

**Pull request merged**
